concent_api/delete-unsupported-messages.py

```diff
@@ -0,0 +1,40 @@
+import django
+import os
+import sys
+
+from logging import getLogger
+
+from common.helpers import deserialize_message
+
+from golem_messages.exceptions import MessageError
+
+sys.path.append('concent_api')
+os.environ['DJANGO_SETTINGS_MODULE'] = "concent_api.settings"
+django.setup()
+
+from core.models import StoredMessage  # noqa E402
+from core.models import Subtask  # noqa E402
+
+logger = getLogger(__name__)
+
+
+def main() -> None:
+    for subtask in Subtask.objects.all():
+        try:
+            report_computed_task = deserialize_message(subtask.report_computed_task.data.tobytes())
+            if not isinstance(report_computed_task.size, int):
+                delete_unsupported_messages(subtask)
+        except MessageError as golem_messages_exception:
+            logger.info(f'During message deserialization exception raised: {golem_messages_exception}')
+            delete_unsupported_messages(subtask)
+
+
+def delete_unsupported_messages(subtask: Subtask) -> None:
+    subtask_id = subtask.subtask_id
+    StoredMessage.objects.filter(subtask_id=subtask_id).delete()
+    subtask.delete()
```

Comment:
You don't need to do `StoredMessage.objects.filter(subtask_id=subtask_id).delete()` explicitly. All the objects that point at the subtask will be deleted when you delete it (unless it's overriden by the foreign key but it it's is, it's probably for a good reason).

See [QuerySet.delete()](https://docs.djangoproject.com/pl/1.11/ref/models/querysets/#delete)

Reply:
@cameel Important one:

It will not because `StoredMessage` objects don't point to `Subtask`. They are only fields on `Subtask`.

```
In [14]: s = Subtask.objects.get(subtask_id=st.subtask_id)
In [16]: s.task_to_compute
Out[16]: <StoredMessage: StoredMessage #61, type:2002, 2018-04-16 10:54:58.821811+00:00>
In [17]: s.delete()
Out[17]: (2, {'core.PaymentInfo': 0, 'core.PendingResponse': 1, 'core.Subtask': 1})
```

So it is actually opposite - if you remove `StoredMessage`, the related `Subtask` gets removed. Is that the behavior we want?

```
In [27]: st
Out[27]: <StoredMessage: StoredMessage #71, type:2002, 2018-04-16 11:03:40.500564+00:00>
In [28]: st.delete()
Out[28]: 
(4,
 {'core.PaymentInfo': 0,
  'core.PendingResponse': 2,
  'core.StoredMessage': 1,
  'core.Subtask': 1})
```

Reply:
@cameel @kbeker 
>  is there a way to change this without changing the model, e.g. via a parameter to delete()?

Yes, by overriding `delete()` method on model and disconnecting related messages before calling `super().delete()`. BUT if we want to do this only for this purpose the same code can be placed here.